### PR TITLE
Make Pool scaling factors fixed-point

### PR DESCRIPTION
**Pull request by nventuro**
Part of https://github.com/balancer-labs/balancer-v2-monorepo/issues/655

### CODE REVIEW

**Comment by nventuro** [comment]
According to the benchmarks this increases gas costs by a bit, but not a large amount. ~200 gas in the worst case (4 token Stable Pool), typically less..

File: pkg/pool-utils/contracts/BasePool.sol [hunk 1]
```diff
@@ -502,14 +502,23 @@ abstract contract BasePool is IBasePool, BasePoolAuthorization, BalancerPoolToke
...
     }
 
     /**
      * @dev Returns the scaling factor for one of the Pool's tokens. Reverts if `token` is not a token registered by the
      * Pool.
+     *
+     * All scaling factors are fixed-point values with 18 decimals, to allow for this function to be overriden by
+     * derived contracts that need to apply further scaling, making these factors potentially non-integer.
+     *
+     * The largest 'base' scaling factor (i.e. in tokens with less than 18 decimals) is 10**18, which in fixed-poin is
```
Comment (TomAFrench) [inline pkg/pool-utils/contracts/BasePool.sol:515]:
```suggestion
     * The largest 'base' scaling factor (i.e. in tokens with less than 18 decimals) is 10**18, which in fixed-point is
```

File: pkg/pool-utils/contracts/BasePool.sol [hunk 2]
```diff
@@ -502,14 +502,23 @@ abstract contract BasePool is IBasePool, BasePoolAuthorization, BalancerPoolToke
...
         uint256 decimalsDifference = Math.sub(18, tokenDecimals);
-        return 10**decimalsDifference;
+        return FixedPoint.ONE * 10**decimalsDifference;
     }
 
     /**
      * @dev Returns the scaling factor for one of the Pool's tokens. Reverts if `token` is not a token registered by the
      * Pool.
+     *
+     * All scaling factors are fixed-point values with 18 decimals, to allow for this function to be overriden by
```
Comment (TomAFrench) [inline pkg/pool-utils/contracts/BasePool.sol:512]:
```suggestion
     * All scaling factors are fixed-point values with 18 decimals, to allow for this function to be overridden by
```

**Review by facuspagnuolo — APPROVED** [review]
LGTM!

File: pkg/pool-utils/contracts/BasePool.sol [hunk 3]
```diff
@@ -561,7 +574,7 @@ abstract contract BasePool is IBasePool, BasePoolAuthorization, BalancerPoolToke
      */
     function _upscaleArray(uint256[] memory amounts, uint256[] memory scalingFactors) internal view {
         for (uint256 i = 0; i < _getTotalTokens(); ++i) {
-            amounts[i] = Math.mul(amounts[i], scalingFactors[i]);
+            amounts[i] = FixedPoint.mulDown(amounts[i], scalingFactors[i]);
```
Comment (dmf7z) [inline pkg/pool-utils/contracts/BasePool.sol:577]: I would add a note here of why we round down in the meantime. Later we can do a bit of more analysis if we need upscaleArrayUp/Down.

**Review by dmf7z — APPROVED** [review]

File: pkg/pool-utils/contracts/BasePool.sol (hunk 3 above)
Comment (nventuro) [reply pkg/pool-utils/contracts/BasePool.sol:577]: I left a note in `_upscale`, since this function is the same except on an array.

**Pull request merged**